File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.py
# Last updated: 6/11/2026, 11:32:43 AM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # appr-1
    # def height(self, root):
    #     if root is None:
    #         return 0
    #     return 1+ max(self.height(root.left), self.height(root.right))
    
    # def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
    #     if(root is None):
    #         return 0
    #     lh = self.diameterOfBinaryTree(root.left)
    #     rh = self.diameterOfBinaryTree(root.right)
    #     ch = self.height(root.left)+self.height(root.right)
    #     return max(lh, rh, ch)

    #appr-2
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ans = [0]
        def diam(root, ans):
            if root is None:
                return 0
            lh = diam(root.left, ans)
            rh= diam(root.right, ans)
            ans[0] = max(ans[0], lh+rh)
            return 1+max(lh, rh)
        diam(root, ans)
        return ans[0]

File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.cpp
// Last updated: 6/11/2026, 11:32:36 AM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxd = 0;
    int helper(TreeNode* root){
        if(root == nullptr){
            return 0;
        }
        int lh = 0;
        if(root->left){
            lh = 1+helper(root->left);
        }
        
        int rh = 0;
        if(root->right){
            rh = 1+helper(root->right);
        }

        maxd = max({maxd, lh, rh, lh+rh});

        // if(root->left == nullptr && root->right == nullptr){
        //     return 1;
        // }

        return max(lh, rh);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        helper(root);
        return maxd;
    }

    
};